src/libstd/sys/common/io.rs

```diff
@@ -0,0 +1,151 @@
+// Copyright 2015 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+use prelude::v1::*;
+use io;
+use io::ErrorKind;
+use io::Read;
+use slice::from_raw_parts_mut;
+
+// Provides read_to_end functionality over an uninitialized buffer.
+// This function is unsafe because it calls the underlying
+// read function with a slice into uninitialized memory. The default
+// implementation of read_to_end for readers will zero out new memory in
+// the buf before passing it to read, but avoiding this zero can often
+// lead to a fairly significant performance win.
+//
+// Implementations using this method have to adhere to two guarantees:
+//  *  The implementation of read never reads the buffer provided.
+//  *  The implementation of read correctly reports how many bytes were written.
+pub unsafe fn read_to_end_uninitialized<R: Read + ?Sized>(mut r: &mut R, buf: &mut Vec<u8>)
+                                                          -> io::Result<usize> {
+
+    let start_len = buf.len();
+    let mut len = start_len;
+    let mut increase =
+        if len == 0 {
+            1
+        } else {
+            len
+        };
+    let ret;
+    loop {
+        if len == buf.capacity() {
+            buf.reserve(increase);
+        }
+
+        let buf_slice = from_raw_parts_mut(buf.as_mut_ptr().offset(len as isize),
+                                           buf.capacity() - len);
+
+        match r.read(buf_slice) {
+            Ok(0) => {
+                ret = Ok(len - start_len);
+                break;
+            }
+            Ok(n) => {
+                len += n;
+                buf.set_len(len);
+                increase = len;
+            },
+            Err(ref e) if e.kind() == ErrorKind::Interrupted => { }
+            Err(e) => {
```

Comment:
This can be condensed a bit now that the final `truncate` isn't needed:

```rust
match r.read(buf) {
    Ok(0) => return Ok(buf.len() - start_len),
    Ok(n) => buf.set_len(buf.len() + n),
    Err(ref e) if e.kind() == ErrorKind::Interrupted => {}
    Err(e) => return Err(e),
}       
```